pkg/services/ngalert/sender/router.go

```diff
@@ -0,0 +1,268 @@
+		d.logger.Info("creating new sender for the external alertmanagers", "org", cfg.OrgID, "alertmanagers", cfg.Alertmanagers)
+		s, err := New()
+		if err != nil {
+			d.logger.Error("unable to start the sender", "err", err, "org", cfg.OrgID)
+			continue
+		}
+
+		d.Senders[cfg.OrgID] = s
+		s.Run()
+
+		err = s.ApplyConfig(cfg)
+		if err != nil {
+			d.logger.Error("failed to apply configuration", "err", err, "org", cfg.OrgID)
+			continue
+		}
+
+		d.SendersCfgHash[cfg.OrgID] = cfg.AsSHA256()
+	}
+
+	sendersToStop := map[int64]*Sender{}
+
+	for orgID, s := range d.Senders {
+		if _, exists := orgsFound[orgID]; !exists {
+			sendersToStop[orgID] = s
+			delete(d.Senders, orgID)
+			delete(d.SendersCfgHash, orgID)
+		}
+	}
+	d.AdminConfigMtx.Unlock()
+
+	// We can now stop these senders w/o having to hold a lock.
+	for orgID, s := range sendersToStop {
+		d.logger.Info("stopping sender", "org", orgID)
+		s.Stop()
+		d.logger.Info("stopped sender", "org", orgID)
+	}
+
+	d.logger.Debug("finish of admin configuration sync")
+
+	return nil
+}
+
+func (d *AlertsRouter) Send(key models.AlertRuleKey, alerts definitions.PostableAlerts) error {
+	logger := d.logger.New("rule_uid", key.UID, "org", key.OrgID)
+	if len(alerts.PostableAlerts) == 0 {
+		logger.Debug("no alerts to notify about")
+		return nil
+	}
+	// Send alerts to local notifier if they need to be handled internally
+	// or if no external AMs have been discovered yet.
+	var localNotifierExist, externalNotifierExist bool
+	if d.SendAlertsTo[key.OrgID] == models.ExternalAlertmanagers && len(d.AlertmanagersFor(key.OrgID)) > 0 {
+		logger.Debug("no alerts to put in the notifier")
+	} else {
+		logger.Debug("sending alerts to local notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
+		n, err := d.MultiOrgNotifier.AlertmanagerFor(key.OrgID)
+		if err == nil {
+			localNotifierExist = true
+			if err := n.PutAlerts(alerts); err != nil {
+				logger.Error("failed to put alerts in the local notifier", "count", len(alerts.PostableAlerts), "err", err)
+			}
+		} else {
+			if errors.Is(err, notifier.ErrNoAlertmanagerForOrg) {
+				logger.Debug("local notifier was not found")
+			} else {
+				logger.Error("local notifier is not available", "err", err)
+			}
+		}
+	}
+
+	// Send alerts to external Alertmanager(s) if we have a sender for this organization
+	// and alerts are not being handled just internally.
+	d.AdminConfigMtx.RLock()
+	defer d.AdminConfigMtx.RUnlock()
+	s, ok := d.Senders[key.OrgID]
+	if ok && d.SendAlertsTo[key.OrgID] != models.InternalAlertmanager {
+		logger.Debug("sending alerts to external notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
+		s.SendAlerts(alerts)
+		externalNotifierExist = true
+	}
+
+	if !localNotifierExist && !externalNotifierExist {
+		return fmt.Errorf("no external or internal notifier - [%d] alerts not delivered", len(alerts.PostableAlerts))
```

Comment:
The caller is not doing anything with this error - so I see no reason to return it (I believe this is one of the functions introduced by this refactor so we can modify it). Can we just log the error here and not return anything from this function.

Reply:
I would like to keep the error here because it is logged by the scheduler that have much more context in the logger (the tick, version of the alert rule, rule key

Reply:
Can you please show an except of how these log lines look like? I think we'll notice that we have enough information as it is from the logs that this is outputting e.g. on error you'll log twice from two different places.

Reply:
I took a look again and it was just that single error that is returned. So, it does not make sense to handle just that one considering the fact that this is an edge case. I changed the interface